File: core/world/world.js
import { canvas, ctx, resizeCanvas } from "./canvas.js";
import { clamp } from "../utils/math.js";
import { toroidalDelta, wrap } from "./utils.js";
import { worldSize } from "../utils/constants.js";

export default class World {
    constructor({ x, y, width, height, scale = 1, object }) {
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
        this.scale = scale;
        this.object = object;
    }

    render() {
        resizeCanvas(this.scale);
        ctx.translate((canvas.width - this.width) / 2, (canvas.height - this.height) / 2);
    }

    update() {
        this.x = wrap(this.x + toroidalDelta(this.x, this.object.x, this.width) * 0.15, this.width);
        this.y = wrap(this.y + toroidalDelta(this.y, this.object.y, this.height) * 0.15, this.height);


        this.angle = this.object.angle;
    }

    zoom(f = 1) {
        this.scale = this.scale * f;
    }

    attach(obj) {
        this.object = obj ?? this.object;
    }
}


export const world = new World({
    x: 0,
    y: 0,
    width: worldSize.width,
    height: worldSize.height,
    scale: 0.7,
})

File: void_front/game/core/world/world.py
from __future__ import annotations

from dataclasses import dataclass

from game.core.utils.constants import worldSize
from .canvas import resizeCanvas


@dataclass
class World:
    x: float
    y: float
    width: float
    height: float
    scale: float = 1.0
    object: object | None = None

    def render(self):
        resizeCanvas(1.0)

    def update(self):
        if self.object is None:
            return
        from .utils import toroidalDelta, wrap

        self.x = wrap(self.x + toroidalDelta(self.x, self.object.x, self.width) * 0.15, self.width)
        self.y = wrap(self.y + toroidalDelta(self.y, self.object.y, self.height) * 0.15, self.height)
        self.angle = getattr(self.object, "angle", 0)

    def zoom(self, f: float = 1.0):
        self.scale *= f

    def attach(self, obj):
        if obj is not None:
            self.object = obj


world = World(x=0, y=0, width=worldSize.width, height=worldSize.height, scale=0.7)